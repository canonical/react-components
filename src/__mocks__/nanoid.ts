let id = 0;

beforeEach(() => {
  id = 0;
});

export const nanoid = (): string => {
  id++;
  return `mock-nanoid-${id}`;
};
