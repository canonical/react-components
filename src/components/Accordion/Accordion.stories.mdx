import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { useState } from "react";
import Accordion from "./Accordion";

<Meta
  title="Accordion"
  component={Accordion}
  argTypes={{
    sections: {
      control: {
        disable: true,
      },
    },
    externallyControlled: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => <Accordion {...args} />;

### Accordion

This is a [React](https://reactjs.org/) component for the Vanilla [Accordion](https://docs.vanillaframework.io/patterns/accordion/).

The sidebar accordion, used in listing pages or as navigation, can hold multiple navigation items or to be used as a filter for content.

### Props

<ArgsTable of={Accordion} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      sections: [
        {
          title: "Advanced topics",
          content: (
            <>
              <p>Charm bundles</p>
              <p>Machine authentication</p>
              <p>Migrating models</p>
              <p>Using storage</p>
              <p>Working with actions</p>
              <p>Working with resources</p>
              <p>Cloud image metadata</p>
              <p>Tools</p>
            </>
          ),
        },
        {
          title: "Networking",
          content: (
            <>
              <p>Working offline</p>
              <p>Fan container networking</p>
              <p>Network spaces</p>
            </>
          ),
        },
        {
          title: "Miscellaneous",
          content: (
            <>
              <p>Juju GUI</p>
              <p>CentOS support</p>
              <p>Collecting Juju metrics</p>
            </>
          ),
        },
      ],
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### External state

The expanded accordion section can be controlled by external state.

<Canvas>
  <Story name="External state">
    {() => {
      const [expandedSection, setExpandedSection] = useState();
      return (
        <Accordion
          expanded={expandedSection}
          externallyControlled
          onExpandedChange={setExpandedSection}
          sections={[
            {
              title: "Advanced topics",
              key: "advanced-topics",
              content: (
                <>
                  <p>Charm bundles</p>
                  <p>Machine authentication</p>
                  <p>Migrating models</p>
                  <p>Using storage</p>
                  <p>Working with actions</p>
                  <p>Working with resources</p>
                  <p>Cloud image metadata</p>
                  <p>Tools</p>
                </>
              ),
            },
            {
              title: "Networking",
              key: "networking",
              content: (
                <>
                  <p>Working offline</p>
                  <p>Fan container networking</p>
                  <p>Network spaces</p>
                </>
              ),
            },
            {
              title: "Miscellaneous",
              key: "miscellaneous",
              content: (
                <>
                  <p>Juju GUI</p>
                  <p>CentOS support</p>
                  <p>Collecting Juju metrics</p>
                </>
              ),
            },
          ]}
        />
      );
    }}
  </Story>
</Canvas>

### Headings

`titleElement` prop can be used to define heading element for section titles.

<Canvas>
  <Story
    name="Headings"
    args={{
      sections: [
        {
          title: "Advanced topics",
          content: (
            <>
              <p>Charm bundles</p>
              <p>Machine authentication</p>
              <p>Migrating models</p>
              <p>Using storage</p>
              <p>Working with actions</p>
              <p>Working with resources</p>
              <p>Cloud image metadata</p>
              <p>Tools</p>
            </>
          ),
        },
        {
          title: "Networking",
          content: (
            <>
              <p>Working offline</p>
              <p>Fan container networking</p>
              <p>Network spaces</p>
            </>
          ),
        },
        {
          title: "Miscellaneous",
          content: (
            <>
              <p>Juju GUI</p>
              <p>CentOS support</p>
              <p>Collecting Juju metrics</p>
            </>
          ),
        },
      ],
      titleElement: "h3",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Custom headings

The `title` prop can be a `ReactNode`, allowing a higher degree of customisation of the section titles.

<Canvas>
  <Story
    name="Custom headings"
    args={{
      sections: [
        {
          title: <>Advanced topics <span className="u-text--muted p-text--small">optional</span></>,
          content: (
            <>
              <p>Charm bundles</p>
              <p>Machine authentication</p>
              <p>Migrating models</p>
              <p>Using storage</p>
              <p>Working with actions</p>
              <p>Working with resources</p>
              <p>Cloud image metadata</p>
              <p>Tools</p>
            </>
          ),
        },
        {
          title: <>Networking <span className="u-text--muted p-text--small">optional</span></>,
          content: (
            <>
              <p>Working offline</p>
              <p>Fan container networking</p>
              <p>Network spaces</p>
            </>
          ),
        },
        {
          title: <>Miscellaneous <span className="u-text--muted p-text--small">optional</span></>,
          content: (
            <>
              <p>Juju GUI</p>
              <p>CentOS support</p>
              <p>Collecting Juju metrics</p>
            </>
          ),
        },
      ],
      titleElement: "h3",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
