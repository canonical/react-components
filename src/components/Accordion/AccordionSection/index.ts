export { default } from "./AccordionSection";
export type {
  Props as AccordionSectionProps,
  AccordionHeadings,
} from "./AccordionSection";
