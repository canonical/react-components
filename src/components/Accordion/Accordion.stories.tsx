import { useState } from "react";
import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Accordion from "./Accordion";

const meta: Meta<typeof Accordion> = {
  component: Accordion,
  tags: ["autodocs"],

  argTypes: {
    sections: {
      control: {
        disable: true,
      },
    },

    externallyControlled: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Accordion>;

export const Default: Story = {
  name: "Default",

  args: {
    sections: [
      {
        title: "Advanced topics",

        content: (
          <>
            <p>Charm bundles</p>
            <p>Machine authentication</p>
            <p>Migrating models</p>
            <p>Using storage</p>
            <p>Working with actions</p>
            <p>Working with resources</p>
            <p>Cloud image metadata</p>
            <p>Tools</p>
          </>
        ),
      },
      {
        title: "Networking",

        content: (
          <>
            <p>Working offline</p>
            <p>Fan container networking</p>
            <p>Network spaces</p>
          </>
        ),
      },
      {
        title: "Miscellaneous",

        content: (
          <>
            <p>Juju GUI</p>
            <p>CentOS support</p>
            <p>Collecting Juju metrics</p>
          </>
        ),
      },
    ],
  },
};

/**
 * The expanded accordion section can be controlled by external state.
 */
export const ExternalState: Story = {
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [expandedSection, setExpandedSection] = useState<string>();

    return (
      <Accordion
        expanded={expandedSection}
        externallyControlled
        onExpandedChange={setExpandedSection}
        sections={[
          {
            title: "Advanced topics",
            key: "advanced-topics",

            content: (
              <>
                <p>Charm bundles</p>
                <p>Machine authentication</p>
                <p>Migrating models</p>
                <p>Using storage</p>
                <p>Working with actions</p>
                <p>Working with resources</p>
                <p>Cloud image metadata</p>
                <p>Tools</p>
              </>
            ),
          },
          {
            title: "Networking",
            key: "networking",

            content: (
              <>
                <p>Working offline</p>
                <p>Fan container networking</p>
                <p>Network spaces</p>
              </>
            ),
          },
          {
            title: "Miscellaneous",
            key: "miscellaneous",

            content: (
              <>
                <p>Juju GUI</p>
                <p>CentOS support</p>
                <p>Collecting Juju metrics</p>
              </>
            ),
          },
        ]}
      />
    );
  },

  name: "External state",
};

/**
 * `titleElement` prop can be used to define heading element for section titles.
 */
export const Headings: Story = {
  name: "Headings",

  args: {
    sections: [
      {
        title: "Advanced topics",

        content: (
          <>
            <p>Charm bundles</p>
            <p>Machine authentication</p>
            <p>Migrating models</p>
            <p>Using storage</p>
            <p>Working with actions</p>
            <p>Working with resources</p>
            <p>Cloud image metadata</p>
            <p>Tools</p>
          </>
        ),
      },
      {
        title: "Networking",

        content: (
          <>
            <p>Working offline</p>
            <p>Fan container networking</p>
            <p>Network spaces</p>
          </>
        ),
      },
      {
        title: "Miscellaneous",

        content: (
          <>
            <p>Juju GUI</p>
            <p>CentOS support</p>
            <p>Collecting Juju metrics</p>
          </>
        ),
      },
    ],

    titleElement: "h3",
  },
};

/**
 * The `title` prop can be a `ReactNode`, allowing a higher degree of customisation of the section titles.
 */
export const CustomHeadings: Story = {
  args: {
    sections: [
      {
        title: (
          <>
            Advanced topics{" "}
            <span className="u-text--muted p-text--small">optional</span>
          </>
        ),
        content: (
          <>
            <p>Charm bundles</p>
            <p>Machine authentication</p>
            <p>Migrating models</p>
            <p>Using storage</p>
            <p>Working with actions</p>
            <p>Working with resources</p>
            <p>Cloud image metadata</p>
            <p>Tools</p>
          </>
        ),
      },
      {
        title: (
          <>
            Networking{" "}
            <span className="u-text--muted p-text--small">optional</span>
          </>
        ),
        content: (
          <>
            <p>Working offline</p>
            <p>Fan container networking</p>
            <p>Network spaces</p>
          </>
        ),
      },
      {
        title: (
          <>
            Miscellaneous{" "}
            <span className="u-text--muted p-text--small">optional</span>
          </>
        ),
        content: (
          <>
            <p>Juju GUI</p>
            <p>CentOS support</p>
            <p>Collecting Juju metrics</p>
          </>
        ),
      },
    ],
    titleElement: "h3",
  },
};
