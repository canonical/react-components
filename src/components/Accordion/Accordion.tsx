import classNames from "classnames";
import React, { useState, HTMLProps } from "react";

import AccordionSection from "./AccordionSection";
import type {
  AccordionHeadings,
  AccordionSectionProps,
} from "./AccordionSection";

import type { ClassName, PropsWithSpread } from "types";

export type Section = AccordionSectionProps & {
  /**
   * An optional key for the section component. It will also be
   * used to track which section is selected.
   */
  key?: string | number;
};

export type Props = PropsWithSpread<
  {
    /**
     * Optional classes applied to the parent element.
     */
    className?: ClassName;
    /**
     * An optional value to set the expanded section. The value must match a
     * section key. This value will only set the expanded section on first render
     * if externallyControlled is not set to `true`.
     */
    expanded?: string;
    /**
     * Whether the expanded section will be controlled via external state.
     */
    externallyControlled?: boolean;
    /**
     * Optional function that is called when the expanded section is changed.
     * The function is provided the section title or null.
     */
    onExpandedChange?: (id: string, title: string) => void;
    /**
     * An array of sections and content.
     */
    sections: Section[];
    /**
     * Optional string describing heading element that should be used for the section titles.
     */
    titleElement?: AccordionHeadings;
  },
  HTMLProps<HTMLElement>
>;

const generateSections = (
  sections: Section[],
  setExpanded,
  expanded,
  titleElement
) =>
  sections.map(({ key, ...props }, i) => (
    <AccordionSection
      expanded={expanded}
      key={key || i}
      sectionKey={key?.toString()}
      setExpanded={setExpanded}
      titleElement={titleElement}
      {...props}
    />
  ));

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Accordion](https://docs.vanillaframework.io/patterns/accordion/).
 *
 * The sidebar accordion, used in listing pages or as navigation, can hold multiple navigation items or to be used as a filter for content.
 */
const Accordion = ({
  className,
  expanded,
  externallyControlled,
  onExpandedChange,
  sections,
  titleElement,
  ...asideProps
}: Props): JSX.Element => {
  const [expandedSection, setExpandedSection] = useState(expanded);
  const setExpanded = (id: string, title: string) => {
    setExpandedSection(id);
    onExpandedChange && onExpandedChange(id, title);
  };
  return (
    <aside
      className={classNames(className, "p-accordion")}
      {...asideProps}
      role="tablist"
      aria-multiselectable="true"
    >
      <ul className="p-accordion__list">
        {generateSections(
          sections,
          setExpanded,
          externallyControlled ? expanded : expandedSection,
          titleElement
        )}
      </ul>
    </aside>
  );
};

export default Accordion;
