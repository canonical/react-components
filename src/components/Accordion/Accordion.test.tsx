import React from "react";
import * as nanoid from "nanoid";

import Accordion from "./Accordion";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Accordion", () => {
  beforeEach(() => {
    jest.spyOn(nanoid, "nanoid").mockReturnValueOnce("mocked-nanoid");
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders", () => {
    render(
      <Accordion
        sections={[
          {
            title: "Advanced topics",
            content: "test content",
          },
          {
            title: "Networking",
            content: <>More test content</>,
          },
        ]}
      />
    );
    expect(screen.getByRole("tablist")).toMatchSnapshot();
  });

  it("passes title heading element to AccordionSections", () => {
    render(
      <Accordion
        sections={[
          {
            title: "Advanced topics",
            content: "test content",
          },
          {
            title: "Networking",
            content: <>More test content</>,
          },
        ]}
        titleElement="h4"
      />
    );
    // Query for the specific element as defined in the titleElement prop.
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector("li h4")).toBeInTheDocument();
  });

  it("can call a function when a section is expanded", async () => {
    const onExpandedChange = jest.fn();
    render(
      <Accordion
        onExpandedChange={onExpandedChange}
        sections={[
          {
            title: "Advanced topics",
            content: "test content",
          },
          {
            title: "Networking",
            content: <>More test content</>,
          },
        ]}
      />
    );
    const tab = screen.getByRole("tab", { name: "Advanced topics" });
    await userEvent.click(tab);
    expect(onExpandedChange).toHaveBeenCalledWith(
      "mocked-nanoid",
      "Advanced topics"
    );
    // Clicking the tab again should close the accordion section.
    await userEvent.click(tab);
    expect(onExpandedChange).toHaveBeenCalledWith(null, null);
  });

  it("can set the default expanded state", () => {
    render(
      <Accordion
        expanded="networking"
        sections={[
          {
            title: "Advanced topics",
            content: "test content",
            key: "advanced",
          },
          {
            title: "Networking",
            content: <>More test content</>,
            key: "networking",
          },
        ]}
      />
    );
    expect(screen.getByRole("tabpanel", { name: "Networking" })).toBeVisible();
  });

  it("can add additional classes", () => {
    render(<Accordion className="extra-class" sections={[]} />);
    expect(screen.getByRole("tablist")).toHaveClass("p-accordion");
    expect(screen.getByRole("tablist")).toHaveClass("extra-class");
  });
});
