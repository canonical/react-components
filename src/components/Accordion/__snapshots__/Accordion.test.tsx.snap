// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders 1`] = `
<aside
  aria-multiselectable="true"
  class="p-accordion"
  role="tablist"
>
  <ul
    class="p-accordion__list"
  >
    <li
      class="p-accordion__group"
    >
      <div
        aria-level="3"
        class="p-accordion__heading"
        role="heading"
      >
        <button
          aria-controls="#mocked-nanoid"
          aria-expanded="false"
          class="p-accordion__tab"
          id="mock-nanoid-1"
          role="tab"
          type="button"
        >
          Advanced topics
        </button>
      </div>
      <section
        aria-hidden="true"
        aria-labelledby="mock-nanoid-1"
        class="p-accordion__panel"
        id="mocked-nanoid"
        role="tabpanel"
      >
        test content
      </section>
    </li>
    <li
      class="p-accordion__group"
    >
      <div
        aria-level="3"
        class="p-accordion__heading"
        role="heading"
      >
        <button
          aria-controls="#mock-nanoid-2"
          aria-expanded="false"
          class="p-accordion__tab"
          id="mock-nanoid-3"
          role="tab"
          type="button"
        >
          Networking
        </button>
      </div>
      <section
        aria-hidden="true"
        aria-labelledby="mock-nanoid-3"
        class="p-accordion__panel"
        id="mock-nanoid-2"
        role="tabpanel"
      >
        More test content
      </section>
    </li>
  </ul>
</aside>
`;
